File: java/src/main/java/ai/guard8/shield/GroupEncryption.java
package ai.guard8.shield;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.*;

/**
 * GroupEncryption - Multi-recipient encryption.
 *
 * Encrypt once for multiple recipients, each can decrypt with their own key.
 * Uses a group key for message encryption, then encrypts the group key
 * separately for each member.
 */
public class GroupEncryption {
    private static final int NONCE_SIZE = 16;
    private static final int MAC_SIZE = 16;
    private static final SecureRandom random = new SecureRandom();

    private byte[] groupKey;
    private final Map<String, byte[]> members = new HashMap<>();

    /**
     * Create group encryption with generated group key.
     */
    public GroupEncryption() {
        this.groupKey = randomBytes(32);
    }

    /**
     * Create group encryption with specified group key.
     */
    public GroupEncryption(byte[] groupKey) {
        if (groupKey.length != 32) {
            throw new IllegalArgumentException("Group key must be 32 bytes");
        }
        this.groupKey = Arrays.copyOf(groupKey, 32);
    }

    /**
     * Add a member to the group.
     *
     * @param memberId Unique member identifier
     * @param sharedKey Pre-shared key with this member (32 bytes)
     */
    public void addMember(String memberId, byte[] sharedKey) {
        if (sharedKey.length != 32) {
            throw new IllegalArgumentException("Shared key must be 32 bytes");
        }
        members.put(memberId, Arrays.copyOf(sharedKey, 32));
    }

    /**
     * Remove a member from the group.
     * Note: After removing a member, you should rotate the group key.
     */
    public boolean removeMember(String memberId) {
        return members.remove(memberId) != null;
    }

    /**
     * Get list of member IDs.
     */
    public List<String> getMembers() {
        return new ArrayList<>(members.keySet());
    }

    /**
     * Encrypt for all group members.
     *
     * @param plaintext Message to encrypt
     * @return Map containing ciphertext and per-member encrypted keys
     */
    public Map<String, Object> encrypt(byte[] plaintext) {
        // Encrypt message with group key
        byte[] ciphertext = encryptBlock(groupKey, plaintext);

        // Encrypt group key for each member
        Map<String, String> encryptedKeys = new HashMap<>();
        for (Map.Entry<String, byte[]> entry : members.entrySet()) {
            byte[] encKey = encryptBlock(entry.getValue(), groupKey);
            encryptedKeys.put(entry.getKey(), Base64.getEncoder().encodeToString(encKey));
        }

        Map<String, Object> result = new HashMap<>();
        result.put("version", 1);
        result.put("ciphertext", Base64.getEncoder().encodeToString(ciphertext));
        result.put("keys", encryptedKeys);
        return result;
    }

    /**
     * Decrypt as a group member.
     *
     * @param encrypted Encrypted message from encrypt()
     * @param memberId Your member ID
     * @param memberKey Your shared key
     * @return Decrypted message, or null if decryption fails
     */
    @SuppressWarnings("unchecked")
    public static byte[] decrypt(Map<String, Object> encrypted, String memberId, byte[] memberKey) {
        Map<String, String> keys = (Map<String, String>) encrypted.get("keys");
        if (keys == null || !keys.containsKey(memberId)) {
            return null;
        }

        // Decrypt group key
        byte[] encryptedGroupKey = Base64.getDecoder().decode(keys.get(memberId));
        byte[] groupKeyDecrypted = decryptBlock(memberKey, encryptedGroupKey);
        if (groupKeyDecrypted == null) {
            return null;
        }

        // Decrypt message
        byte[] ciphertext = Base64.getDecoder().decode((String) encrypted.get("ciphertext"));
        return decryptBlock(groupKeyDecrypted, ciphertext);
    }

    /**
     * Rotate the group key.
     *
     * @return The old group key
     */
    public byte[] rotateKey() {
        byte[] oldKey = groupKey;
        groupKey = randomBytes(32);
        return oldKey;
    }

    // ============== Helper Methods ==============

    private static byte[] encryptBlock(byte[] key, byte[] data) {
        byte[] nonce = randomBytes(NONCE_SIZE);
        byte[] keystream = generateKeystream(key, nonce, data.length);
        byte[] ciphertext = new byte[data.length];
        for (int i = 0; i < data.length; i++) {
            ciphertext[i] = (byte) (data[i] ^ keystream[i]);
        }

        byte[] macData = new byte[NONCE_SIZE + ciphertext.length];
        System.arraycopy(nonce, 0, macData, 0, NONCE_SIZE);
        System.arraycopy(ciphertext, 0, macData, NONCE_SIZE, ciphertext.length);
        byte[] mac = hmacSha256(key, macData);

        byte[] result = new byte[NONCE_SIZE + ciphertext.length + MAC_SIZE];
        System.arraycopy(nonce, 0, result, 0, NONCE_SIZE);
        System.arraycopy(ciphertext, 0, result, NONCE_SIZE, ciphertext.length);
        System.arraycopy(mac, 0, result, NONCE_SIZE + ciphertext.length, MAC_SIZE);

        return result;
    }

    private static byte[] decryptBlock(byte[] key, byte[] encrypted) {
        if (encrypted.length < NONCE_SIZE + MAC_SIZE) {
            return null;
        }

        byte[] nonce = Arrays.copyOfRange(encrypted, 0, NONCE_SIZE);
        byte[] ciphertext = Arrays.copyOfRange(encrypted, NONCE_SIZE, encrypted.length - MAC_SIZE);
        byte[] receivedMac = Arrays.copyOfRange(encrypted, encrypted.length - MAC_SIZE, encrypted.length);

        byte[] macData = new byte[NONCE_SIZE + ciphertext.length];
        System.arraycopy(nonce, 0, macData, 0, NONCE_SIZE);
        System.arraycopy(ciphertext, 0, macData, NONCE_SIZE, ciphertext.length);
        byte[] expectedMac = hmacSha256(key, macData);

        if (!constantTimeEquals(receivedMac, Arrays.copyOf(expectedMac, MAC_SIZE))) {
            return null;
        }

        byte[] keystream = generateKeystream(key, nonce, ciphertext.length);
        byte[] decrypted = new byte[ciphertext.length];
        for (int i = 0; i < ciphertext.length; i++) {
            decrypted[i] = (byte) (ciphertext[i] ^ keystream[i]);
        }

        return decrypted;
    }

    private static byte[] generateKeystream(byte[] key, byte[] nonce, int length) {
        int numBlocks = (length + 31) / 32;
        byte[] keystream = new byte[numBlocks * 32];

        for (int i = 0; i < numBlocks; i++) {
            ByteBuffer buf = ByteBuffer.allocate(32 + NONCE_SIZE + 4).order(ByteOrder.LITTLE_ENDIAN);
            buf.put(key);
            buf.put(nonce);
            buf.putInt(i);
            byte[] hash = sha256(buf.array());
            System.arraycopy(hash, 0, keystream, i * 32, 32);
        }

        return Arrays.copyOf(keystream, length);
    }

    private static byte[] sha256(byte[] data) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return md.digest(data);
        } catch (Exception e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    private static byte[] hmacSha256(byte[] key, byte[] data) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(key, "HmacSHA256"));
            return mac.doFinal(data);
        } catch (Exception e) {
            throw new RuntimeException("HMAC-SHA256 not available", e);
        }
    }

    private static boolean constantTimeEquals(byte[] a, byte[] b) {
        if (a.length != b.length) return false;
        int result = 0;
        for (int i = 0; i < a.length; i++) {
            result |= a[i] ^ b[i];
        }
        return result == 0;
    }

    private static byte[] randomBytes(int length) {
        byte[] bytes = new byte[length];
        random.nextBytes(bytes);
        return bytes;
    }
}

File: kotlin/src/main/kotlin/ai/guard8/shield/GroupEncryption.kt
package ai.guard8.shield

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * GroupEncryption - Multi-recipient encryption.
 *
 * Encrypt once for multiple recipients, each can decrypt with their own key.
 * Uses a group key for message encryption, then encrypts the group key
 * separately for each member.
 */
class GroupEncryption private constructor(private var groupKey: ByteArray) : AutoCloseable {

    private val members = mutableMapOf<String, ByteArray>()

    companion object {
        private const val NONCE_SIZE = 16
        private const val MAC_SIZE = 16
        private val random = SecureRandom()

        /**
         * Create group encryption with generated group key.
         */
        fun create(): GroupEncryption {
            return GroupEncryption(randomBytes(32))
        }

        /**
         * Create group encryption with specified group key.
         */
        fun create(groupKey: ByteArray): GroupEncryption {
            require(groupKey.size == 32) { "Group key must be 32 bytes" }
            return GroupEncryption(groupKey.copyOf())
        }

        /**
         * Decrypt as a group member.
         */
        @Suppress("UNCHECKED_CAST")
        fun decrypt(encrypted: Map<String, Any>, memberId: String, memberKey: ByteArray): ByteArray? {
            val keys = encrypted["keys"] as? Map<String, String> ?: return null
            val encryptedKeyB64 = keys[memberId] ?: return null

            // Decrypt group key
            val encryptedGroupKey = Base64.getDecoder().decode(encryptedKeyB64)
            val groupKey = decryptBlock(memberKey, encryptedGroupKey) ?: return null

            // Decrypt message
            val ciphertext = Base64.getDecoder().decode(encrypted["ciphertext"] as String)
            return decryptBlock(groupKey, ciphertext)
        }

        private fun encryptBlock(key: ByteArray, data: ByteArray): ByteArray {
            val nonce = randomBytes(NONCE_SIZE)
            val keystream = generateKeystream(key, nonce, data.size)
            val ciphertext = ByteArray(data.size)
            for (i in data.indices) {
                ciphertext[i] = (data[i].toInt() xor keystream[i].toInt()).toByte()
            }

            val macData = nonce + ciphertext
            val mac = hmacSha256(key, macData)

            return nonce + ciphertext + mac.copyOf(MAC_SIZE)
        }

        private fun decryptBlock(key: ByteArray, encrypted: ByteArray): ByteArray? {
            if (encrypted.size < NONCE_SIZE + MAC_SIZE) return null

            val nonce = encrypted.copyOfRange(0, NONCE_SIZE)
            val ciphertext = encrypted.copyOfRange(NONCE_SIZE, encrypted.size - MAC_SIZE)
            val receivedMac = encrypted.copyOfRange(encrypted.size - MAC_SIZE, encrypted.size)

            val macData = nonce + ciphertext
            val expectedMac = hmacSha256(key, macData).copyOf(MAC_SIZE)

            if (!constantTimeEquals(receivedMac, expectedMac)) return null

            val keystream = generateKeystream(key, nonce, ciphertext.size)
            val decrypted = ByteArray(ciphertext.size)
            for (i in ciphertext.indices) {
                decrypted[i] = (ciphertext[i].toInt() xor keystream[i].toInt()).toByte()
            }

            return decrypted
        }

        private fun generateKeystream(key: ByteArray, nonce: ByteArray, length: Int): ByteArray {
            val numBlocks = (length + 31) / 32
            val keystream = ByteArray(numBlocks * 32)

            for (i in 0 until numBlocks) {
                val block = ByteBuffer.allocate(32 + NONCE_SIZE + 4).order(ByteOrder.LITTLE_ENDIAN)
                block.put(key)
                block.put(nonce)
                block.putInt(i)
                val hash = sha256(block.array())
                System.arraycopy(hash, 0, keystream, i * 32, 32)
            }

            return keystream.copyOf(length)
        }

        private fun sha256(data: ByteArray): ByteArray {
            return MessageDigest.getInstance("SHA-256").digest(data)
        }

        private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(key, "HmacSHA256"))
            return mac.doFinal(data)
        }

        private fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean {
            if (a.size != b.size) return false
            var result = 0
            for (i in a.indices) {
                result = result or (a[i].toInt() xor b[i].toInt())
            }
            return result == 0
        }

        private fun randomBytes(length: Int): ByteArray {
            val bytes = ByteArray(length)
            random.nextBytes(bytes)
            return bytes
        }
    }

    /**
     * Add a member to the group.
     */
    fun addMember(memberId: String, sharedKey: ByteArray) {
        require(sharedKey.size == 32) { "Shared key must be 32 bytes" }
        members[memberId] = sharedKey.copyOf()
    }

    /**
     * Remove a member from the group.
     */
    fun removeMember(memberId: String): Boolean {
        return members.remove(memberId) != null
    }

    /**
     * Get list of member IDs.
     */
    fun getMembers(): List<String> = members.keys.toList()

    /**
     * Encrypt for all group members.
     */
    fun encrypt(plaintext: ByteArray): Map<String, Any> {
        // Encrypt message with group key
        val ciphertext = encryptBlock(groupKey, plaintext)

        // Encrypt group key for each member
        val encryptedKeys = mutableMapOf<String, String>()
        for ((memberId, memberKey) in members) {
            val encKey = encryptBlock(memberKey, groupKey)
            encryptedKeys[memberId] = Base64.getEncoder().encodeToString(encKey)
        }

        return mapOf(
            "version" to 1,
            "ciphertext" to Base64.getEncoder().encodeToString(ciphertext),
            "keys" to encryptedKeys
        )
    }

    /**
     * Rotate the group key.
     */
    fun rotateKey(): ByteArray {
        val oldKey = groupKey
        groupKey = randomBytes(32)
        return oldKey
    }

    override fun close() {
        groupKey.fill(0)
        members.values.forEach { it.fill(0) }
        members.clear()
    }
}
